Public/Src/Engine/Processes/AsyncProcessExecutor.cs

```diff
@@ -0,0 +1,235 @@
+
+        /// <summary>
+        /// Checks if process exit is completed.
+        /// </summary>
+        public bool ExitCompleted => m_processExitedTcs.Task.IsCompleted;
+
+        /// <summary>
+        /// Checks if standard out flush is completed.
+        /// </summary>
+        public bool StdOutCompleted => m_stdoutFlushedTcs.Task.IsCompleted;
+
+        /// <summary>
+        /// Checks if standard error flush is completed.
+        /// </summary>
+        public bool StdErrCompleted => m_stderrFlushedTcs.Task.IsCompleted;
+
+        /// <summary>
+        /// Timeout.
+        /// </summary>
+        private readonly TimeSpan m_timeout;
+
+        /// <summary>
+        /// Sandboxed process info for provenence purpose.
+        /// </summary>
+        private readonly SandboxedProcessInfo m_sandboxedProcessInfo;
+
+        private int m_processId = -1;
+
+        /// <summary>
+        /// Process id.
+        /// </summary>
+        public int ProcessId => m_processId != -1 ? m_processId : (m_processId = Process.Id);
+
+        /// <summary>
+        /// Gets active peak memory usage.
+        /// </summary>
+        public ulong? GetActivePeakMemoryUsage()
+        {
+            try
+            {
+                if (Process == null || Process.HasExited)
+                {
+                    return null;
+                }
+
+                return Dispatch.GetActivePeakMemoryUsage(Process.Handle, ProcessId);
+            }
+#pragma warning disable ERP022 // Unobserved exception in generic exception handler
+            catch
+            {
+                return null;
+            }
+#pragma warning restore ERP022 // Unobserved exception in generic exception handler
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="AsyncProcessExecutor"/>.
+        /// </summary>
+        public AsyncProcessExecutor(
+            Process process, 
+            TimeSpan timeout,
+            Action<string> outputBuilder = null,
+            Action<string> errorBuilder = null,
+            SandboxedProcessInfo sandboxedProcessInfo = null)
+        {
+            Contract.Requires(process != null);
+
+            Process = process;
+            Process.Exited += (sender, e) => m_processExitedTcs.TrySetResult(Unit.Void);
+
+            if (outputBuilder != null)
+            {
+                process.OutputDataReceived += (sender, e) => FeedOutputBuilder(m_stdoutFlushedTcs, e.Data, outputBuilder);
+            }
+
+            if (errorBuilder != null)
+            {
+                process.ErrorDataReceived += (sender, e) => FeedOutputBuilder(m_stderrFlushedTcs, e.Data, errorBuilder);
+            }
+
+            m_timeout = timeout;
+            m_sandboxedProcessInfo = sandboxedProcessInfo;
+        }
+
+        /// <summary>
+        /// Starts process.
+        /// </summary>
+        public void Start()
+        {
+            if (!FileUtilities.FileExistsNoFollow(Process.StartInfo.FileName))

```

Comment:
Any particular reason to check `FileExistsNoFollow` here?  Do you intentionally want to forbid running a process through a symlink?

Reply:
There's no particular reason. It's a kind a reflex after working on symlinks for some time :-)

---
In reply to: [277334192](https://github.com/Microsoft/BuildXL/pull/159#discussion_r277334192) [](ancestors = 277334192)